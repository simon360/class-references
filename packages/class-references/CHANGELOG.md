# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="0.1.0"></a>
# [0.1.0](https://github.com/simon360/class-references/compare/v0.0.4...v0.1.0) (2018-08-25)


### Features

* **class-references:** rename functions for clarity ([674fb87](https://github.com/simon360/class-references/commit/674fb87))





<a name="0.0.4"></a>
## [0.0.4](https://github.com/simon360/class-references/compare/v0.0.3...v0.0.4) (2018-08-25)

**Note:** Version bump only for package class-references





<a name="0.0.3"></a>
## [0.0.3](https://github.com/simon360/class-references/compare/v0.0.2...v0.0.3) (2018-08-25)

**Note:** Version bump only for package class-references





<a name="0.0.2"></a>
## [0.0.2](https://github.com/simon360/class-references/compare/v0.0.1...v0.0.2) (2018-08-25)

**Note:** Version bump only for package class-references
