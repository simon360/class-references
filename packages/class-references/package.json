{
  "name": "class-references",
  "version": "0.0.0-pre",
  "description": "Add or remove a class using reference counting",
  "main": "lib/index.js",
  "author": "Simon Andrews <me@simonandrews.ca>",
  "license": "MIT",
  "scripts": {
    "compile": "tsc"
  }
}
