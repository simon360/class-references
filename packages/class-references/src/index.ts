import { getClassReferences, setClassReferences } from "./element";
import {
  newObjectWithoutReferenceToId,
  newObjectWithReferenceToClass
} from "./referenceObject";

/**
 * Add {className} to {el}, if it wasn't previously added, and return a unique
 * token for the transaction.
 *
 * @param el the element to add a class to
 * @param className the class to add
 */
export const claimForClass = (el: HTMLElement, className: string) => {
  const newReferences = newObjectWithReferenceToClass(
    getClassReferences(el),
    className
  );

  el.classList.add(className);
  setClassReferences(el, newReferences);

  return newReferences.lastToken;
};

/**
 *
 * @param el the element to release a reference on
 * @param token the token to release
 */
export const release = (el: HTMLElement, token: number) => {
  const newReferences = newObjectWithoutReferenceToId(
    getClassReferences(el),
    token
  );

  // `newObjectWithoutReferenceToId` won't immediately remove empty lists
  // from `classes`.
  for (const className of Object.keys(
    Object.entries(newReferences.classes)
      .filter(([_, values]) => values.length === 0)
      .map(([key]) => key)
  )) {
    el.classList.remove(className);
  }

  setClassReferences(el, newReferences);
};
