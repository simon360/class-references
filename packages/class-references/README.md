# `class-references`

> A framework-agnostic library for managing multiple references to class names.

For more details, see the [main README file](https://github.com/simon360/class-references/blob/master/README.md).

## Installation

Using npm:

```
npm install --save class-references
```

Using yarn:

```
yarn add class-references
```

## Usage

Any time you want to add a class (eg. `MyClassName`) to an element (eg. `document.body`), use:

```js
import { requestTokenForClass } from "class-references";

const token = requestTokenForClass(document.body, "MyClassName");
```

When you're ready to remove the class, make sure you can recall the token, and use:

```js
import { releaseToken } from "class-references";

releaseToken(document.body, token);
```

### Advanced examples

```js
import { releaseToken, requestTokenForClass } from "class-references";

const className = "u-preventScroll";
const element = document.body;

const token1 = requestTokenForClass(element, className); // Class is added
const token2 = requestTokenForClass(element, className); // Class was already added, but another token reserves it.

releaseToken(element, token1); // token1 is released; token2 not released, class remains
releaseToken(element, token2); // both tokens are released, class removed
```
