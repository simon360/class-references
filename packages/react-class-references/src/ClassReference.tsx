import { claimForClass, release } from "class-references";
import * as React from "react";

interface IProps {
  children: any;
  classNameToAdd: string;
  element: HTMLElement;
}

interface IState {
  token: number;
}

export default class ClassReference extends React.Component<IProps, IState> {
  public componentDidMount() {
    this.setState({
      token: claimForClass(this.props.element, this.props.classNameToAdd)
    });
  }

  public componentDidUpdate(prevProps: IProps) {
    const { classNameToAdd, element } = this.props;
    const { token } = this.state;

    if (
      prevProps.element !== element ||
      prevProps.classNameToAdd !== classNameToAdd
    ) {
      release(prevProps.element, token);

      this.setState({
        token: claimForClass(element, classNameToAdd)
      });
    }
  }

  public componentWillUnmount() {
    release(this.props.element, this.state.token);
  }

  public render() {
    return <React.Fragment>{this.props.children}</React.Fragment>;
  }
}
