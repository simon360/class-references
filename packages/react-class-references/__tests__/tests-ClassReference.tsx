import { claimForClass, release } from "class-references";
import * as React from "react";
import * as TestRenderer from "react-test-renderer";
import ClassReference from "../src/ClassReference";

jest.mock("class-references");

describe("<ClassReference />", () => {
  const testClassName = "test-class-name";
  const fakeToken = 12;
  let element: HTMLElement;

  beforeEach(() => {
    element = document.createElement("div");
    jest.resetAllMocks();

    (claimForClass as any).mockReturnValue(fakeToken);
  });

  it("should match default snapshot", () => {
    expect(
      TestRenderer.create(
        <ClassReference element={element} classNameToAdd={testClassName}>
          Hello, world!
        </ClassReference>
      ).toJSON()
    ).toMatchSnapshot();
  });

  it("should call claimForClass", () => {
    TestRenderer.create(
      <ClassReference element={element} classNameToAdd={testClassName}>
        Hello, world
      </ClassReference>
    );

    expect(claimForClass).toHaveBeenCalledWith(element, testClassName);
  });

  it("should release on unmount", () => {
    const renderer = TestRenderer.create(
      <ClassReference element={element} classNameToAdd={testClassName}>
        Hello, world
      </ClassReference>
    );

    renderer.unmount();

    expect(release).toHaveBeenCalledWith(element, fakeToken);
  });

  it("should release and claim on element swap", () => {
    const extraElement = document.createElement("div");

    const Test = ({ testElement }: { testElement: HTMLElement }) => (
      <ClassReference element={testElement} classNameToAdd={testClassName}>
        Hello, world
      </ClassReference>
    );

    const renderer = TestRenderer.create(<Test testElement={element} />);

    renderer.update(<Test testElement={extraElement} />);

    expect(release).toHaveBeenCalledWith(element, fakeToken);
    expect(claimForClass).toHaveBeenCalledWith(extraElement, testClassName);
  });

  it("should release and claim on class swap", () => {
    const extraClassName = "a-different-class-name";

    const Test = ({ classNameToAdd }: { classNameToAdd: string }) => (
      <ClassReference element={element} classNameToAdd={classNameToAdd}>
        Hello, world
      </ClassReference>
    );

    const renderer = TestRenderer.create(
      <Test classNameToAdd={testClassName} />
    );

    renderer.update(<Test classNameToAdd={extraClassName} />);

    expect(release).toHaveBeenCalledWith(element, fakeToken);
    expect(claimForClass).toHaveBeenCalledWith(element, extraClassName);
  });
});
