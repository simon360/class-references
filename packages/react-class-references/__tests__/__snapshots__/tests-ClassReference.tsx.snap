// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClassReference /> should match default snapshot 1`] = `"Hello, world!"`;
